File: algo/generate-parentheses/generateParenthesis.go
func add(result *[]string, str string, m, n int) {
    if m == 0 && n == 0 {
        *result = append(*result, str)
    } else {
        if n > 0 {
            add(result, str + ")", m, n - 1)
        }
        if m > 0 {
            add(result, str + "(", m - 1, n + 1)
        }
    }
}

func generateParenthesis(n int) []string {
    l := &[]string{}
    add(l, "", n, 0)
    return *l
}

File: algo/generate-parentheses/generateParenthesis.py
class Solution(object):
    def add(self, l, s, n, m):
        if n == 0 and m == 0:
            l.append(s)
        else:
            if m > 0: self.add(l, s + ')', n, m - 1)
            if n > 0: self.add(l, s + '(', n - 1, m + 1)
    
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        l = []
        self.add(l, '', n, 0)
        return l